{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. General info of dataset GSE198896\n",
    "\n",
    "This is the Jupyter Notebook for dataset GSE198896. Its dataset includes a .tar.gz file for each sample; within each tar file, there is a sample directory including barcodes & features.tsv,  matrix mtx files.\n",
    "\n",
    "Thus, we need to first gzip and then incorparate these barcodes/genes/matrix files and generate an AnnData object for each sample. In total, there are 18 samples.\n",
    "\n",
    "<span style=\"color:green\">**[PBMC]**</span> Peripheral blood mononuclear cell\n",
    "\n",
    "<span style=\"color:green\">**[TIL]**</span> Tumor-Infiltrating Lymphocyte\n",
    "\n",
    "<span style=\"color:green\">**[CITE-SEQ]**</span> Simultaneously quantifies cell surface protein and transcriptomic data within a single cell readout"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Environment setup\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import anndata as anndata\n",
    "import scipy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. AnnData object of each sample\n",
    "\n",
    "<strike><span style=\"color:red\">**IMPORTANT:**</span> gzip all the files</strike> --> see Problem 1 below\n",
    "\n",
    "1. `barcodes.tsv.gz`: cell barcodes, which go into `.obs`\n",
    "2. `features.tsv.gz`: gene names, `.var`\n",
    "3. `matrix.mtx.gz`: the expression matrix, `.X`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 3854 × 33538\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids', 'feature_types'\n",
      "GSM5959134_Osteosarcoma_2_TIL\n",
      "GSM5959144_Ewings_3_CITEseq\n",
      "GSM5959136_Osteosarcoma_3_TIL\n",
      "GSM5959138_Osteosarcoma_4_TIL\n",
      "GSM5959133_Osteosarcoma_2_PBMC\n",
      "GSM5959145_Ewings_4_CITEseq\n",
      "AnnData object with n_obs × n_vars = 3390 × 33538\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids', 'feature_types'\n",
      "GSM5959135_Osteosarcoma_3_PBMC\n",
      "GSM5959146_Healthy_donor_CITEseq\n",
      "AnnData object with n_obs × n_vars = 10843 × 33538\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids', 'feature_types'\n",
      "GSM5959129_Ewing_1_PBMC\n",
      "GSM5959130_Ewing_1_TIL\n",
      "AnnData object with n_obs × n_vars = 4296 × 33538\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids', 'feature_types'\n",
      "GSM5959132_Osteosarcoma_1_TIL\n",
      "GSM5959137_Osteosarcoma_4_PBMC\n",
      "GSM5959131_Osteosarcoma_1_PBMC\n",
      "GSM5959143_Ewings_2_CITEseq\n"
     ]
    }
   ],
   "source": [
    "# from pathlib import Path\n",
    "\n",
    "# # Specify directory paths\n",
    "# data_directory = Path('/scratch/user/s4543064/xiaohan-john-project/data/GSE198896')\n",
    "# write_directory = Path('/scratch/user/s4543064/xiaohan-john-project/write/GSE198896')\n",
    "\n",
    "# # Loop through all files in the directory\n",
    "# for sample_directory in data_directory.iterdir():\n",
    "#     sample_name = sample_directory.stem\n",
    "#     sample_h5ad = sample_name + '_uni.h5ad'\n",
    "\n",
    "#     try:\n",
    "#         sample = sc.read_10x_mtx(\n",
    "#         sample_directory,\n",
    "#         var_names='gene_symbols',  \n",
    "#         cache=False\n",
    "#         )\n",
    "\n",
    "#         # Create an observation metric info to store related features\n",
    "#         obs_metrics = pd.DataFrame(index=sample.obs_names) ## Get the identifiers\n",
    "#         if \"Ewing\" in sample_name:\n",
    "#             obs_metrics['cancer_type'] = 'ewing_sarcoma'\n",
    "#         else:\n",
    "#             obs_metrics['cancer_type'] = 'osteosarcoma'\n",
    "        \n",
    "#         obs_metrics['dataset'] = 'GSE198896'\n",
    "        \n",
    "#         if \"PBMC\" in sample_name:\n",
    "#             obs_metrics['tissue'] = 'pbmc'\n",
    "#         elif \"TIL\" in sample_name:\n",
    "#             obs_metrics['tissue'] = 'sarcoma_tumour'\n",
    "        \n",
    "#         obs_metrics['sample_barcode'] = sample_name\n",
    "#         obs_metrics['uni_barcode'] = obs_metrics['dataset'] + '_' + obs_metrics.index.astype(str)\n",
    "        \n",
    "#         sample.obs = obs_metrics\n",
    "#         sample.obs.set_index(\"uni_barcode\", drop=False, inplace=True)\n",
    "#         print(sample)\n",
    "\n",
    "#         # save the anndata object\n",
    "#         output_path = write_directory / sample_h5ad\n",
    "#         sample.write_h5ad(output_path, compression=\"gzip\")\n",
    "#     except:\n",
    "#         print(sample_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style=\"color:red\">**PROBLEM 1:**</span> The following samples have incomplete features.tsv.gz files <br>\n",
    "`GSM5959134_Osteosarcoma_2_TIL`<br>\n",
    "`GSM5959136_Osteosarcoma_3_TIL`<br>\n",
    "`GSM5959138_Osteosarcoma_4_TIL`<br>\n",
    "`GSM5959133_Osteosarcoma_2_PBMC`<br>\n",
    "`GSM5959135_Osteosarcoma_3_PBMC`<br>\n",
    "`GSM5959129_Ewing_1_PBMC`<br>\n",
    "`GSM5959130_Ewing_1_TIL`<br>\n",
    "`GSM5959132_Osteosarcoma_1_TIL`<br>\n",
    "`GSM5959137_Osteosarcoma_4_PBMC`<br>\n",
    "`GSM5959131_Osteosarcoma_1_PBMC`<br>\n",
    "We expect to have three columns in the features.tsv.gz files (1st col: gene identifier, 2nd col: gene symbol, 3rd col: feature type), but these samples lack the 3rd column.\n",
    "\n",
    "So the easiest way is to unzip all barcodes/features/matrix files and rename features.tsv to genes.tsv --> treat them as outputs from older version of cell ranger.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 3854 × 33538\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'disease_progression', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 687 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'disease_progression', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "GSM5959144_Ewings_3_CITEseq\n",
      "AnnData object with n_obs × n_vars = 3495 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'disease_progression', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 2008 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'disease_progression', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 1498 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'disease_progression', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "GSM5959145_Ewings_4_CITEseq\n",
      "AnnData object with n_obs × n_vars = 3390 × 33538\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'disease_progression', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 1693 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'disease_progression', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "GSM5959146_Healthy_donor_CITEseq\n",
      "AnnData object with n_obs × n_vars = 10843 × 33538\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'disease_progression', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 1631 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'disease_progression', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 2376 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'disease_progression', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 4296 × 33538\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'disease_progression', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 1690 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'disease_progression', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 572 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'disease_progression', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 2655 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'disease_progression', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "GSM5959143_Ewings_2_CITEseq\n"
     ]
    }
   ],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "# Specify directory paths\n",
    "data_directory = Path('/scratch/user/s4543064/xiaohan-john-project/data/GSE198896')\n",
    "write_directory = Path('/scratch/user/s4543064/xiaohan-john-project/write/GSE198896')\n",
    "\n",
    "# Loop through all files in the directory\n",
    "for sample_directory in data_directory.iterdir():\n",
    "    sample_name = sample_directory.stem\n",
    "    sample_h5ad = sample_name + '_uni.h5ad'\n",
    "\n",
    "    try:\n",
    "        sample = sc.read_10x_mtx(\n",
    "        sample_directory,\n",
    "        var_names='gene_symbols',  \n",
    "        cache=False\n",
    "        )\n",
    "\n",
    "        # Create an observation metric info to store related features\n",
    "        obs_metrics = pd.DataFrame(index=sample.obs_names) ## Get the identifiers\n",
    "        if \"Ewing\" in sample_name:\n",
    "            obs_metrics['cancer_type'] = 'ewing_sarcoma'\n",
    "        else:\n",
    "            obs_metrics['cancer_type'] = 'osteosarcoma'\n",
    "        \n",
    "        obs_metrics['dataset'] = 'GSE198896'\n",
    "        \n",
    "        if \"PBMC\" in sample_name:\n",
    "            obs_metrics['tissue'] = 'pbmc'\n",
    "        elif \"TIL\" in sample_name:\n",
    "            obs_metrics['tissue'] = 'bone/surrounding_soft_tissue'\n",
    "        \n",
    "        if \"Healthy\" in sample_name:\n",
    "            obs_metrics['disease_progression'] = 'healthy'\n",
    "        else:\n",
    "            obs_metrics['disease_progression'] = obs_metrics['cancer_type']\n",
    "        \n",
    "        obs_metrics['sample_barcode'] = sample_name\n",
    "        obs_metrics['uni_barcode'] = obs_metrics['dataset'] + '_' + obs_metrics.index.astype(str)\n",
    "        \n",
    "        sample.obs = obs_metrics\n",
    "        sample.obs.set_index(\"uni_barcode\", drop=False, inplace=True)\n",
    "        print(sample)\n",
    "\n",
    "        # save the anndata object\n",
    "        output_path = write_directory / sample_h5ad\n",
    "        sample.write_h5ad(output_path, compression=\"gzip\")\n",
    "    except:\n",
    "        print(sample_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cancer_type</th>\n",
       "      <th>dataset</th>\n",
       "      <th>tissue</th>\n",
       "      <th>disease_progression</th>\n",
       "      <th>sample_barcode</th>\n",
       "      <th>uni_barcode</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>uni_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSE198896_AAACCTGAGCCAGAAC-1</th>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSE198896</td>\n",
       "      <td>pbmc</td>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSM5959131_Osteosarcoma_1_PBMC</td>\n",
       "      <td>GSE198896_AAACCTGAGCCAGAAC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE198896_AAACCTGAGCGAGAAA-1</th>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSE198896</td>\n",
       "      <td>pbmc</td>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSM5959131_Osteosarcoma_1_PBMC</td>\n",
       "      <td>GSE198896_AAACCTGAGCGAGAAA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE198896_AAACCTGAGGTGCAAC-1</th>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSE198896</td>\n",
       "      <td>pbmc</td>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSM5959131_Osteosarcoma_1_PBMC</td>\n",
       "      <td>GSE198896_AAACCTGAGGTGCAAC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE198896_AAACCTGAGTACACCT-1</th>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSE198896</td>\n",
       "      <td>pbmc</td>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSM5959131_Osteosarcoma_1_PBMC</td>\n",
       "      <td>GSE198896_AAACCTGAGTACACCT-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE198896_AAACCTGCACCGAAAG-1</th>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSE198896</td>\n",
       "      <td>pbmc</td>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSM5959131_Osteosarcoma_1_PBMC</td>\n",
       "      <td>GSE198896_AAACCTGCACCGAAAG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE198896_TTTGTCAAGACTCGGA-1</th>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSE198896</td>\n",
       "      <td>pbmc</td>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSM5959131_Osteosarcoma_1_PBMC</td>\n",
       "      <td>GSE198896_TTTGTCAAGACTCGGA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE198896_TTTGTCAAGGTCGGAT-1</th>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSE198896</td>\n",
       "      <td>pbmc</td>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSM5959131_Osteosarcoma_1_PBMC</td>\n",
       "      <td>GSE198896_TTTGTCAAGGTCGGAT-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE198896_TTTGTCACATGACGGA-1</th>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSE198896</td>\n",
       "      <td>pbmc</td>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSM5959131_Osteosarcoma_1_PBMC</td>\n",
       "      <td>GSE198896_TTTGTCACATGACGGA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE198896_TTTGTCATCCAGATCA-1</th>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSE198896</td>\n",
       "      <td>pbmc</td>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSM5959131_Osteosarcoma_1_PBMC</td>\n",
       "      <td>GSE198896_TTTGTCATCCAGATCA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE198896_TTTGTCATCTACCAGA-1</th>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSE198896</td>\n",
       "      <td>pbmc</td>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSM5959131_Osteosarcoma_1_PBMC</td>\n",
       "      <td>GSE198896_TTTGTCATCTACCAGA-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2655 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               cancer_type    dataset tissue  \\\n",
       "uni_barcode                                                    \n",
       "GSE198896_AAACCTGAGCCAGAAC-1  osteosarcoma  GSE198896   pbmc   \n",
       "GSE198896_AAACCTGAGCGAGAAA-1  osteosarcoma  GSE198896   pbmc   \n",
       "GSE198896_AAACCTGAGGTGCAAC-1  osteosarcoma  GSE198896   pbmc   \n",
       "GSE198896_AAACCTGAGTACACCT-1  osteosarcoma  GSE198896   pbmc   \n",
       "GSE198896_AAACCTGCACCGAAAG-1  osteosarcoma  GSE198896   pbmc   \n",
       "...                                    ...        ...    ...   \n",
       "GSE198896_TTTGTCAAGACTCGGA-1  osteosarcoma  GSE198896   pbmc   \n",
       "GSE198896_TTTGTCAAGGTCGGAT-1  osteosarcoma  GSE198896   pbmc   \n",
       "GSE198896_TTTGTCACATGACGGA-1  osteosarcoma  GSE198896   pbmc   \n",
       "GSE198896_TTTGTCATCCAGATCA-1  osteosarcoma  GSE198896   pbmc   \n",
       "GSE198896_TTTGTCATCTACCAGA-1  osteosarcoma  GSE198896   pbmc   \n",
       "\n",
       "                             disease_progression  \\\n",
       "uni_barcode                                        \n",
       "GSE198896_AAACCTGAGCCAGAAC-1        osteosarcoma   \n",
       "GSE198896_AAACCTGAGCGAGAAA-1        osteosarcoma   \n",
       "GSE198896_AAACCTGAGGTGCAAC-1        osteosarcoma   \n",
       "GSE198896_AAACCTGAGTACACCT-1        osteosarcoma   \n",
       "GSE198896_AAACCTGCACCGAAAG-1        osteosarcoma   \n",
       "...                                          ...   \n",
       "GSE198896_TTTGTCAAGACTCGGA-1        osteosarcoma   \n",
       "GSE198896_TTTGTCAAGGTCGGAT-1        osteosarcoma   \n",
       "GSE198896_TTTGTCACATGACGGA-1        osteosarcoma   \n",
       "GSE198896_TTTGTCATCCAGATCA-1        osteosarcoma   \n",
       "GSE198896_TTTGTCATCTACCAGA-1        osteosarcoma   \n",
       "\n",
       "                                              sample_barcode  \\\n",
       "uni_barcode                                                    \n",
       "GSE198896_AAACCTGAGCCAGAAC-1  GSM5959131_Osteosarcoma_1_PBMC   \n",
       "GSE198896_AAACCTGAGCGAGAAA-1  GSM5959131_Osteosarcoma_1_PBMC   \n",
       "GSE198896_AAACCTGAGGTGCAAC-1  GSM5959131_Osteosarcoma_1_PBMC   \n",
       "GSE198896_AAACCTGAGTACACCT-1  GSM5959131_Osteosarcoma_1_PBMC   \n",
       "GSE198896_AAACCTGCACCGAAAG-1  GSM5959131_Osteosarcoma_1_PBMC   \n",
       "...                                                      ...   \n",
       "GSE198896_TTTGTCAAGACTCGGA-1  GSM5959131_Osteosarcoma_1_PBMC   \n",
       "GSE198896_TTTGTCAAGGTCGGAT-1  GSM5959131_Osteosarcoma_1_PBMC   \n",
       "GSE198896_TTTGTCACATGACGGA-1  GSM5959131_Osteosarcoma_1_PBMC   \n",
       "GSE198896_TTTGTCATCCAGATCA-1  GSM5959131_Osteosarcoma_1_PBMC   \n",
       "GSE198896_TTTGTCATCTACCAGA-1  GSM5959131_Osteosarcoma_1_PBMC   \n",
       "\n",
       "                                               uni_barcode  \n",
       "uni_barcode                                                 \n",
       "GSE198896_AAACCTGAGCCAGAAC-1  GSE198896_AAACCTGAGCCAGAAC-1  \n",
       "GSE198896_AAACCTGAGCGAGAAA-1  GSE198896_AAACCTGAGCGAGAAA-1  \n",
       "GSE198896_AAACCTGAGGTGCAAC-1  GSE198896_AAACCTGAGGTGCAAC-1  \n",
       "GSE198896_AAACCTGAGTACACCT-1  GSE198896_AAACCTGAGTACACCT-1  \n",
       "GSE198896_AAACCTGCACCGAAAG-1  GSE198896_AAACCTGCACCGAAAG-1  \n",
       "...                                                    ...  \n",
       "GSE198896_TTTGTCAAGACTCGGA-1  GSE198896_TTTGTCAAGACTCGGA-1  \n",
       "GSE198896_TTTGTCAAGGTCGGAT-1  GSE198896_TTTGTCAAGGTCGGAT-1  \n",
       "GSE198896_TTTGTCACATGACGGA-1  GSE198896_TTTGTCACATGACGGA-1  \n",
       "GSE198896_TTTGTCATCCAGATCA-1  GSE198896_TTTGTCATCCAGATCA-1  \n",
       "GSE198896_TTTGTCATCTACCAGA-1  GSE198896_TTTGTCATCTACCAGA-1  \n",
       "\n",
       "[2655 rows x 6 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style=\"color:red\">**PROBLEM 2:**</span> The genes.tsv files for all 4 CITEseq samples are incomplete. The standard genes.tsv should have two columns (1st column: gene ID such as ENSG00000186092; 2nd column: gene symbol such as OR4F5). However, the genes,tsv for these samples are like:  \n",
    "\n",
    "hto2-TGATGGCCTATTGGG<br>\n",
    "hto3-TTCCGCCTCTCTTTG<br>\n",
    "hto4-AGTAAGTTCAGCGTA<br>\n",
    "hto5-AAGTATCGTTTCGCA<br>\n",
    "unmapped\n",
    "\n",
    "These look like cellular indices (hashtags) instead of gene-related annotations."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Confirmation of created AnnData objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 2655 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'disease_progression', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 1690 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'disease_progression', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 2376 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'disease_progression', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 1631 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'disease_progression', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 572 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'disease_progression', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 1498 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'disease_progression', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 3390 × 33538\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'disease_progression', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 4296 × 33538\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'disease_progression', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 687 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'disease_progression', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 1693 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'disease_progression', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 10843 × 33538\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'disease_progression', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 3495 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'disease_progression', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 3854 × 33538\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'disease_progression', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 2008 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'disease_progression', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n"
     ]
    }
   ],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "# Specify directory paths\n",
    "write_directory = Path('/scratch/user/s4543064/xiaohan-john-project/write/GSE198896')\n",
    "\n",
    "# Loop through all files in the directory\n",
    "for file in write_directory.iterdir():\n",
    "    if \"_uni.h5ad\" in file.name:\n",
    "        sample = anndata.read_h5ad(file)\n",
    "        print(sample)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cancer_type</th>\n",
       "      <th>dataset</th>\n",
       "      <th>tissue</th>\n",
       "      <th>disease_progression</th>\n",
       "      <th>sample_barcode</th>\n",
       "      <th>uni_barcode</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>uni_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSE198896_AAACCTGGTGTAACGG-1</th>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSE198896</td>\n",
       "      <td>bone/surrounding_soft_tissue</td>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSM5959138_Osteosarcoma_4_TIL</td>\n",
       "      <td>GSE198896_AAACCTGGTGTAACGG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE198896_AAACCTGTCTATCCCG-1</th>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSE198896</td>\n",
       "      <td>bone/surrounding_soft_tissue</td>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSM5959138_Osteosarcoma_4_TIL</td>\n",
       "      <td>GSE198896_AAACCTGTCTATCCCG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE198896_AAACGGGGTAATCACC-1</th>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSE198896</td>\n",
       "      <td>bone/surrounding_soft_tissue</td>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSM5959138_Osteosarcoma_4_TIL</td>\n",
       "      <td>GSE198896_AAACGGGGTAATCACC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE198896_AAACGGGGTTTAGCTG-1</th>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSE198896</td>\n",
       "      <td>bone/surrounding_soft_tissue</td>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSM5959138_Osteosarcoma_4_TIL</td>\n",
       "      <td>GSE198896_AAACGGGGTTTAGCTG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE198896_AAAGATGAGACTAGGC-1</th>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSE198896</td>\n",
       "      <td>bone/surrounding_soft_tissue</td>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSM5959138_Osteosarcoma_4_TIL</td>\n",
       "      <td>GSE198896_AAAGATGAGACTAGGC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE198896_TTTGGTTTCCGAATGT-1</th>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSE198896</td>\n",
       "      <td>bone/surrounding_soft_tissue</td>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSM5959138_Osteosarcoma_4_TIL</td>\n",
       "      <td>GSE198896_TTTGGTTTCCGAATGT-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE198896_TTTGTCAAGACAATAC-1</th>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSE198896</td>\n",
       "      <td>bone/surrounding_soft_tissue</td>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSM5959138_Osteosarcoma_4_TIL</td>\n",
       "      <td>GSE198896_TTTGTCAAGACAATAC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE198896_TTTGTCAGTAACGCGA-1</th>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSE198896</td>\n",
       "      <td>bone/surrounding_soft_tissue</td>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSM5959138_Osteosarcoma_4_TIL</td>\n",
       "      <td>GSE198896_TTTGTCAGTAACGCGA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE198896_TTTGTCAGTAGCCTAT-1</th>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSE198896</td>\n",
       "      <td>bone/surrounding_soft_tissue</td>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSM5959138_Osteosarcoma_4_TIL</td>\n",
       "      <td>GSE198896_TTTGTCAGTAGCCTAT-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE198896_TTTGTCATCGTTACAG-1</th>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSE198896</td>\n",
       "      <td>bone/surrounding_soft_tissue</td>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>GSM5959138_Osteosarcoma_4_TIL</td>\n",
       "      <td>GSE198896_TTTGTCATCGTTACAG-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2008 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               cancer_type    dataset  \\\n",
       "uni_barcode                                             \n",
       "GSE198896_AAACCTGGTGTAACGG-1  osteosarcoma  GSE198896   \n",
       "GSE198896_AAACCTGTCTATCCCG-1  osteosarcoma  GSE198896   \n",
       "GSE198896_AAACGGGGTAATCACC-1  osteosarcoma  GSE198896   \n",
       "GSE198896_AAACGGGGTTTAGCTG-1  osteosarcoma  GSE198896   \n",
       "GSE198896_AAAGATGAGACTAGGC-1  osteosarcoma  GSE198896   \n",
       "...                                    ...        ...   \n",
       "GSE198896_TTTGGTTTCCGAATGT-1  osteosarcoma  GSE198896   \n",
       "GSE198896_TTTGTCAAGACAATAC-1  osteosarcoma  GSE198896   \n",
       "GSE198896_TTTGTCAGTAACGCGA-1  osteosarcoma  GSE198896   \n",
       "GSE198896_TTTGTCAGTAGCCTAT-1  osteosarcoma  GSE198896   \n",
       "GSE198896_TTTGTCATCGTTACAG-1  osteosarcoma  GSE198896   \n",
       "\n",
       "                                                    tissue  \\\n",
       "uni_barcode                                                  \n",
       "GSE198896_AAACCTGGTGTAACGG-1  bone/surrounding_soft_tissue   \n",
       "GSE198896_AAACCTGTCTATCCCG-1  bone/surrounding_soft_tissue   \n",
       "GSE198896_AAACGGGGTAATCACC-1  bone/surrounding_soft_tissue   \n",
       "GSE198896_AAACGGGGTTTAGCTG-1  bone/surrounding_soft_tissue   \n",
       "GSE198896_AAAGATGAGACTAGGC-1  bone/surrounding_soft_tissue   \n",
       "...                                                    ...   \n",
       "GSE198896_TTTGGTTTCCGAATGT-1  bone/surrounding_soft_tissue   \n",
       "GSE198896_TTTGTCAAGACAATAC-1  bone/surrounding_soft_tissue   \n",
       "GSE198896_TTTGTCAGTAACGCGA-1  bone/surrounding_soft_tissue   \n",
       "GSE198896_TTTGTCAGTAGCCTAT-1  bone/surrounding_soft_tissue   \n",
       "GSE198896_TTTGTCATCGTTACAG-1  bone/surrounding_soft_tissue   \n",
       "\n",
       "                             disease_progression  \\\n",
       "uni_barcode                                        \n",
       "GSE198896_AAACCTGGTGTAACGG-1        osteosarcoma   \n",
       "GSE198896_AAACCTGTCTATCCCG-1        osteosarcoma   \n",
       "GSE198896_AAACGGGGTAATCACC-1        osteosarcoma   \n",
       "GSE198896_AAACGGGGTTTAGCTG-1        osteosarcoma   \n",
       "GSE198896_AAAGATGAGACTAGGC-1        osteosarcoma   \n",
       "...                                          ...   \n",
       "GSE198896_TTTGGTTTCCGAATGT-1        osteosarcoma   \n",
       "GSE198896_TTTGTCAAGACAATAC-1        osteosarcoma   \n",
       "GSE198896_TTTGTCAGTAACGCGA-1        osteosarcoma   \n",
       "GSE198896_TTTGTCAGTAGCCTAT-1        osteosarcoma   \n",
       "GSE198896_TTTGTCATCGTTACAG-1        osteosarcoma   \n",
       "\n",
       "                                             sample_barcode  \\\n",
       "uni_barcode                                                   \n",
       "GSE198896_AAACCTGGTGTAACGG-1  GSM5959138_Osteosarcoma_4_TIL   \n",
       "GSE198896_AAACCTGTCTATCCCG-1  GSM5959138_Osteosarcoma_4_TIL   \n",
       "GSE198896_AAACGGGGTAATCACC-1  GSM5959138_Osteosarcoma_4_TIL   \n",
       "GSE198896_AAACGGGGTTTAGCTG-1  GSM5959138_Osteosarcoma_4_TIL   \n",
       "GSE198896_AAAGATGAGACTAGGC-1  GSM5959138_Osteosarcoma_4_TIL   \n",
       "...                                                     ...   \n",
       "GSE198896_TTTGGTTTCCGAATGT-1  GSM5959138_Osteosarcoma_4_TIL   \n",
       "GSE198896_TTTGTCAAGACAATAC-1  GSM5959138_Osteosarcoma_4_TIL   \n",
       "GSE198896_TTTGTCAGTAACGCGA-1  GSM5959138_Osteosarcoma_4_TIL   \n",
       "GSE198896_TTTGTCAGTAGCCTAT-1  GSM5959138_Osteosarcoma_4_TIL   \n",
       "GSE198896_TTTGTCATCGTTACAG-1  GSM5959138_Osteosarcoma_4_TIL   \n",
       "\n",
       "                                               uni_barcode  \n",
       "uni_barcode                                                 \n",
       "GSE198896_AAACCTGGTGTAACGG-1  GSE198896_AAACCTGGTGTAACGG-1  \n",
       "GSE198896_AAACCTGTCTATCCCG-1  GSE198896_AAACCTGTCTATCCCG-1  \n",
       "GSE198896_AAACGGGGTAATCACC-1  GSE198896_AAACGGGGTAATCACC-1  \n",
       "GSE198896_AAACGGGGTTTAGCTG-1  GSE198896_AAACGGGGTTTAGCTG-1  \n",
       "GSE198896_AAAGATGAGACTAGGC-1  GSE198896_AAAGATGAGACTAGGC-1  \n",
       "...                                                    ...  \n",
       "GSE198896_TTTGGTTTCCGAATGT-1  GSE198896_TTTGGTTTCCGAATGT-1  \n",
       "GSE198896_TTTGTCAAGACAATAC-1  GSE198896_TTTGTCAAGACAATAC-1  \n",
       "GSE198896_TTTGTCAGTAACGCGA-1  GSE198896_TTTGTCAGTAACGCGA-1  \n",
       "GSE198896_TTTGTCAGTAGCCTAT-1  GSE198896_TTTGTCAGTAGCCTAT-1  \n",
       "GSE198896_TTTGTCATCGTTACAG-1  GSE198896_TTTGTCATCGTTACAG-1  \n",
       "\n",
       "[2008 rows x 6 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. Convert AnnData objects to SingleCellExperiment objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "class: SingleCellExperiment \n",
      "dim: 33694 2655 \n",
      "metadata(0):\n",
      "assays(1): X\n",
      "rownames(33694): RP11-34P13.3 FAM138A ... AC213203.1 FAM231B\n",
      "rowData names(1): gene_ids\n",
      "colnames(2655): GSE198896_AAACCTGAGCCAGAAC-1\n",
      "  GSE198896_AAACCTGAGCGAGAAA-1 ... GSE198896_TTTGTCATCCAGATCA-1\n",
      "  GSE198896_TTTGTCATCTACCAGA-1\n",
      "colData names(6): cancer_type dataset ... sample_barcode uni_barcode\n",
      "reducedDimNames(0):\n",
      "mainExpName: NULL\n",
      "altExpNames(0):\n",
      "\n",
      "class: SingleCellExperiment \n",
      "dim: 33694 1690 \n",
      "metadata(0):\n",
      "assays(1): X\n",
      "rownames(33694): RP11-34P13.3 FAM138A ... AC213203.1 FAM231B\n",
      "rowData names(1): gene_ids\n",
      "colnames(1690): GSE198896_AAACCTGAGAATTGTG-1\n",
      "  GSE198896_AAACCTGCACATCCGG-1 ... GSE198896_TTTGGTTTCTGCTGCT-1\n",
      "  GSE198896_TTTGTCAAGATGCGAC-1\n",
      "colData names(6): cancer_type dataset ... sample_barcode uni_barcode\n",
      "reducedDimNames(0):\n",
      "mainExpName: NULL\n",
      "altExpNames(0):\n",
      "\n",
      "class: SingleCellExperiment \n",
      "dim: 33694 2376 \n",
      "metadata(0):\n",
      "assays(1): X\n",
      "rownames(33694): RP11-34P13.3 FAM138A ... AC213203.1 FAM231B\n",
      "rowData names(1): gene_ids\n",
      "colnames(2376): GSE198896_AAACCTGCAGGAATCG-1\n",
      "  GSE198896_AAACCTGTCGCTGATA-1 ... GSE198896_TTTGTCAGTCTTCTCG-1\n",
      "  GSE198896_TTTGTCATCTCCTATA-1\n",
      "colData names(6): cancer_type dataset ... sample_barcode uni_barcode\n",
      "reducedDimNames(0):\n",
      "mainExpName: NULL\n",
      "altExpNames(0):\n",
      "\n",
      "class: SingleCellExperiment \n",
      "dim: 33694 1631 \n",
      "metadata(0):\n",
      "assays(1): X\n",
      "rownames(33694): RP11-34P13.3 FAM138A ... AC213203.1 FAM231B\n",
      "rowData names(1): gene_ids\n",
      "colnames(1631): GSE198896_AAACCTGAGGCATGTG-1\n",
      "  GSE198896_AAACCTGAGTAATCCC-1 ... GSE198896_TTTGGTTTCTTCGGTC-1\n",
      "  GSE198896_TTTGTCAAGATCGGGT-1\n",
      "colData names(6): cancer_type dataset ... sample_barcode uni_barcode\n",
      "reducedDimNames(0):\n",
      "mainExpName: NULL\n",
      "altExpNames(0):\n",
      "\n",
      "class: SingleCellExperiment \n",
      "dim: 33694 572 \n",
      "metadata(0):\n",
      "assays(1): X\n",
      "rownames(33694): RP11-34P13.3 FAM138A ... AC213203.1 FAM231B\n",
      "rowData names(1): gene_ids\n",
      "colnames(572): GSE198896_AAACGGGCAAGCCGCT-1\n",
      "  GSE198896_AAACGGGTCCTACAGA-1 ... GSE198896_TTTGTCACATACTACG-1\n",
      "  GSE198896_TTTGTCAGTACAAGTA-1\n",
      "colData names(6): cancer_type dataset ... sample_barcode uni_barcode\n",
      "reducedDimNames(0):\n",
      "mainExpName: NULL\n",
      "altExpNames(0):\n",
      "\n",
      "class: SingleCellExperiment \n",
      "dim: 33694 1498 \n",
      "metadata(0):\n",
      "assays(1): X\n",
      "rownames(33694): RP11-34P13.3 FAM138A ... AC213203.1 FAM231B\n",
      "rowData names(1): gene_ids\n",
      "colnames(1498): GSE198896_AAACCTGAGGTAAACT-1\n",
      "  GSE198896_AAACCTGTCCTGCTTG-1 ... GSE198896_TTTGTCAGTCTCGTTC-1\n",
      "  GSE198896_TTTGTCATCTATCCCG-1\n",
      "colData names(6): cancer_type dataset ... sample_barcode uni_barcode\n",
      "reducedDimNames(0):\n",
      "mainExpName: NULL\n",
      "altExpNames(0):\n",
      "\n",
      "class: SingleCellExperiment \n",
      "dim: 33538 3390 \n",
      "metadata(0):\n",
      "assays(1): X\n",
      "rownames(33538): MIR1302-2HG FAM138A ... AC213203.1 FAM231C\n",
      "rowData names(1): gene_ids\n",
      "colnames(3390): GSE198896_AAACCTGAGGCGTACA-1\n",
      "  GSE198896_AAACCTGAGTGCCATT-1 ... GSE198896_TTTGTCAGTACCTACA-1\n",
      "  GSE198896_TTTGTCAGTGGCAAAC-1\n",
      "colData names(6): cancer_type dataset ... sample_barcode uni_barcode\n",
      "reducedDimNames(0):\n",
      "mainExpName: NULL\n",
      "altExpNames(0):\n",
      "\n",
      "class: SingleCellExperiment \n",
      "dim: 33538 4296 \n",
      "metadata(0):\n",
      "assays(1): X\n",
      "rownames(33538): MIR1302-2HG FAM138A ... AC213203.1 FAM231C\n",
      "rowData names(1): gene_ids\n",
      "colnames(4296): GSE198896_AAACCTGAGCAGGTCA-1\n",
      "  GSE198896_AAACCTGCATCACCCT-1 ... GSE198896_TTTGTCACATGGTTGT-1\n",
      "  GSE198896_TTTGTCATCCAAAGTC-1\n",
      "colData names(6): cancer_type dataset ... sample_barcode uni_barcode\n",
      "reducedDimNames(0):\n",
      "mainExpName: NULL\n",
      "altExpNames(0):\n",
      "\n",
      "class: SingleCellExperiment \n",
      "dim: 33694 687 \n",
      "metadata(0):\n",
      "assays(1): X\n",
      "rownames(33694): RP11-34P13.3 FAM138A ... AC213203.1 FAM231B\n",
      "rowData names(1): gene_ids\n",
      "colnames(687): GSE198896_AAACCTGTCACGCGGT-1\n",
      "  GSE198896_AAACCTGTCTCTGCTG-1 ... GSE198896_TTTGCGCTCATTCACT-1\n",
      "  GSE198896_TTTGGTTCAGATGGCA-1\n",
      "colData names(6): cancer_type dataset ... sample_barcode uni_barcode\n",
      "reducedDimNames(0):\n",
      "mainExpName: NULL\n",
      "altExpNames(0):\n",
      "\n",
      "class: SingleCellExperiment \n",
      "dim: 33694 1693 \n",
      "metadata(0):\n",
      "assays(1): X\n",
      "rownames(33694): RP11-34P13.3 FAM138A ... AC213203.1 FAM231B\n",
      "rowData names(1): gene_ids\n",
      "colnames(1693): GSE198896_AAACCTGAGACCACGA-1\n",
      "  GSE198896_AAACCTGTCGTTGCCT-1 ... GSE198896_TTTGGTTTCCCTCTTT-1\n",
      "  GSE198896_TTTGTCACAAAGTGCG-1\n",
      "colData names(6): cancer_type dataset ... sample_barcode uni_barcode\n",
      "reducedDimNames(0):\n",
      "mainExpName: NULL\n",
      "altExpNames(0):\n",
      "\n",
      "class: SingleCellExperiment \n",
      "dim: 33538 10843 \n",
      "metadata(0):\n",
      "assays(1): X\n",
      "rownames(33538): MIR1302-2HG FAM138A ... AC213203.1 FAM231C\n",
      "rowData names(1): gene_ids\n",
      "colnames(10843): GSE198896_AAACCTGAGACAGAGA-1\n",
      "  GSE198896_AAACCTGAGAGACTTA-1 ... GSE198896_TTTGTCATCGCTAGCG-1\n",
      "  GSE198896_TTTGTCATCTTAACCT-1\n",
      "colData names(6): cancer_type dataset ... sample_barcode uni_barcode\n",
      "reducedDimNames(0):\n",
      "mainExpName: NULL\n",
      "altExpNames(0):\n",
      "\n",
      "class: SingleCellExperiment \n",
      "dim: 33694 3495 \n",
      "metadata(0):\n",
      "assays(1): X\n",
      "rownames(33694): RP11-34P13.3 FAM138A ... AC213203.1 FAM231B\n",
      "rowData names(1): gene_ids\n",
      "colnames(3495): GSE198896_AAACCTGAGCATCATC-1\n",
      "  GSE198896_AAACCTGAGCGGATCA-1 ... GSE198896_TTTGTCATCTCACATT-1\n",
      "  GSE198896_TTTGTCATCTTTACAC-1\n",
      "colData names(6): cancer_type dataset ... sample_barcode uni_barcode\n",
      "reducedDimNames(0):\n",
      "mainExpName: NULL\n",
      "altExpNames(0):\n",
      "\n",
      "class: SingleCellExperiment \n",
      "dim: 33538 3854 \n",
      "metadata(0):\n",
      "assays(1): X\n",
      "rownames(33538): MIR1302-2HG FAM138A ... AC213203.1 FAM231C\n",
      "rowData names(1): gene_ids\n",
      "colnames(3854): GSE198896_AAACCTGAGTACATGA-1\n",
      "  GSE198896_AAACCTGAGTGCAAGC-1 ... GSE198896_TTTGTCAGTAGAAAGG-1\n",
      "  GSE198896_TTTGTCATCTCAAGTG-1\n",
      "colData names(6): cancer_type dataset ... sample_barcode uni_barcode\n",
      "reducedDimNames(0):\n",
      "mainExpName: NULL\n",
      "altExpNames(0):\n",
      "\n",
      "class: SingleCellExperiment \n",
      "dim: 33694 2008 \n",
      "metadata(0):\n",
      "assays(1): X\n",
      "rownames(33694): RP11-34P13.3 FAM138A ... AC213203.1 FAM231B\n",
      "rowData names(1): gene_ids\n",
      "colnames(2008): GSE198896_AAACCTGGTGTAACGG-1\n",
      "  GSE198896_AAACCTGTCTATCCCG-1 ... GSE198896_TTTGTCAGTAGCCTAT-1\n",
      "  GSE198896_TTTGTCATCGTTACAG-1\n",
      "colData names(6): cancer_type dataset ... sample_barcode uni_barcode\n",
      "reducedDimNames(0):\n",
      "mainExpName: NULL\n",
      "altExpNames(0):\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "import anndata2ri\n",
    "import rpy2.robjects as robjects\n",
    "from rpy2.robjects import r\n",
    "from rpy2.robjects.conversion import localconverter\n",
    "\n",
    "# Specify directory paths\n",
    "write_directory = Path('/scratch/user/s4543064/xiaohan-john-project/write/GSE198896')\n",
    "\n",
    "# Loop through all files in the directory\n",
    "for file in write_directory.iterdir():\n",
    "    sample_name = file.stem\n",
    "    if \"_uni.h5ad\" in file.name:\n",
    "        sample_anndata = anndata.read_h5ad(file)\n",
    "        sample_sce_file = sample_name + \".rds\"\n",
    "\n",
    "        with localconverter(anndata2ri.converter):\n",
    "            sample_sce = anndata2ri.py2rpy(sample_anndata)\n",
    "        print(sample_sce)\n",
    "        \n",
    "        # Save the sce object in .rds file\n",
    "        robjects.globalenv[\"sample_sce\"] = sample_sce\n",
    "        sample_sce_path = write_directory / sample_sce_file\n",
    "        robjects.r(\"saveRDS(sample_sce, file='{}')\".format(sample_sce_path))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
