{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. General info of dataset GSE140819\n",
    "\n",
    "This is the Jupyter Notebook for dataset GSE140819. Its dataset includes a single h5 files for each sample.\n",
    ". \n",
    "In total, there are 40 samples from various origins processed with different technologies. Within the pediatric tumours are:\n",
    "\n",
    "Neuroblastoma (NB): HTAPP-312-SMP-901, HTAPP-312-SMP-902, HTAPP-656-SMP-3481, HTAPP-244-SMP-451 (nuclei), HTAPP-656-SMP-3481 (nuclei)\n",
    "\n",
    "Glioblastoma (GB): HTAPP-443-SMP-5491\n",
    "\n",
    "Sarcoma: HTAPP-951-SMP-4652 (nuclei), HTAPP-975-SMP-4771 (nuclei)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/scratch/temp/9167408/ipykernel_1330560/2319647765.py:3: DeprecationWarning: \n",
      "Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),\n",
      "(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)\n",
      "but was not found to be installed on your system.\n",
      "If this would cause problems for you,\n",
      "please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466\n",
      "        \n",
      "  import pandas as pd\n"
     ]
    }
   ],
   "source": [
    "# Environment setup\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import anndata as anndata\n",
    "import scipy\n",
    "\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/s4543064/miniforge3/envs/scvi_env/lib/python3.11/site-packages/anndata/_core/anndata.py:1820: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/home/s4543064/miniforge3/envs/scvi_env/lib/python3.11/site-packages/anndata/_core/anndata.py:1820: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 6794880 × 33538\n",
       "    var: 'gene_ids', 'feature_types', 'genome'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# inspect the provided h5ad file\n",
    "data_directory = Path('/scratch/user/s4543064/xiaohan-john-project/data/GSE140819')\n",
    "\n",
    "adata_path = data_directory / 'GSM4186995_HTAPP-975-SMP-4771_TST-V3_channel1_raw_feature_bc_matrix.h5'\n",
    "adata = sc.read_10x_h5(adata_path)\n",
    "\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>genome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.3</th>\n",
       "      <td>ENSG00000239945</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC233755.2</th>\n",
       "      <td>ENSG00000277856</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC233755.1</th>\n",
       "      <td>ENSG00000275063</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC240274.1</th>\n",
       "      <td>ENSG00000271254</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC213203.1</th>\n",
       "      <td>ENSG00000277475</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM231C</th>\n",
       "      <td>ENSG00000268674</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33538 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    gene_ids    feature_types                genome\n",
       "MIR1302-2HG  ENSG00000243485  Gene Expression  GRCh38-3.0.0_premrna\n",
       "FAM138A      ENSG00000237613  Gene Expression  GRCh38-3.0.0_premrna\n",
       "OR4F5        ENSG00000186092  Gene Expression  GRCh38-3.0.0_premrna\n",
       "AL627309.1   ENSG00000238009  Gene Expression  GRCh38-3.0.0_premrna\n",
       "AL627309.3   ENSG00000239945  Gene Expression  GRCh38-3.0.0_premrna\n",
       "...                      ...              ...                   ...\n",
       "AC233755.2   ENSG00000277856  Gene Expression  GRCh38-3.0.0_premrna\n",
       "AC233755.1   ENSG00000275063  Gene Expression  GRCh38-3.0.0_premrna\n",
       "AC240274.1   ENSG00000271254  Gene Expression  GRCh38-3.0.0_premrna\n",
       "AC213203.1   ENSG00000277475  Gene Expression  GRCh38-3.0.0_premrna\n",
       "FAM231C      ENSG00000268674  Gene Expression  GRCh38-3.0.0_premrna\n",
       "\n",
       "[33538 rows x 3 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAAACACT-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAAACCAT-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAAACCCA-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAAACCCG-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAAACCTG-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGTCTTTGCTA-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGTCTTTGCTG-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGTCTTTGGAG-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGTCTTTGGCT-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGTCTTTGTCG-1</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6794880 rows × 0 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [AAACCCAAGAAACACT-1, AAACCCAAGAAACCAT-1, AAACCCAAGAAACCCA-1, AAACCCAAGAAACCCG-1, AAACCCAAGAAACCTG-1, AAACCCAAGAAACGAA-1, AAACCCAAGAAACGTC-1, AAACCCAAGAAACTAC-1, AAACCCAAGAAACTCA-1, AAACCCAAGAAACTGC-1, AAACCCAAGAAACTGT-1, AAACCCAAGAAAGAAC-1, AAACCCAAGAAAGACA-1, AAACCCAAGAAAGCCT-1, AAACCCAAGAAAGCGA-1, AAACCCAAGAAAGGAT-1, AAACCCAAGAAAGGTA-1, AAACCCAAGAAAGTCT-1, AAACCCAAGAAATAGG-1, AAACCCAAGAAATCCA-1, AAACCCAAGAAATGAG-1, AAACCCAAGAAATGCC-1, AAACCCAAGAAATGGG-1, AAACCCAAGAAATTCG-1, AAACCCAAGAAATTGC-1, AAACCCAAGAACAAGG-1, AAACCCAAGAACACCA-1, AAACCCAAGAACACGT-1, AAACCCAAGAACAGGA-1, AAACCCAAGAACCCGA-1, AAACCCAAGAACCCGG-1, AAACCCAAGAACCGAT-1, AAACCCAAGAACCGCA-1, AAACCCAAGAACCTCC-1, AAACCCAAGAACCTTT-1, AAACCCAAGAACGAGG-1, AAACCCAAGAACGCGT-1, AAACCCAAGAACGCTA-1, AAACCCAAGAACGTGC-1, AAACCCAAGAACGTGT-1, AAACCCAAGAACTACG-1, AAACCCAAGAACTCCT-1, AAACCCAAGAACTCTT-1, AAACCCAAGAACTGAC-1, AAACCCAAGAACTGAT-1, AAACCCAAGAACTGGT-1, AAACCCAAGAACTTCC-1, AAACCCAAGAACTTTC-1, AAACCCAAGAAGAAAC-1, AAACCCAAGAAGAACG-1, AAACCCAAGAAGAATG-1, AAACCCAAGAAGAGCA-1, AAACCCAAGAAGAGGC-1, AAACCCAAGAAGAGTA-1, AAACCCAAGAAGATAG-1, AAACCCAAGAAGATCT-1, AAACCCAAGAAGCCAC-1, AAACCCAAGAAGCCAT-1, AAACCCAAGAAGCCCG-1, AAACCCAAGAAGCCGC-1, AAACCCAAGAAGCCTA-1, AAACCCAAGAAGCCTG-1, AAACCCAAGAAGCGAA-1, AAACCCAAGAAGCGCT-1, AAACCCAAGAAGCGGG-1, AAACCCAAGAAGCGTT-1, AAACCCAAGAAGCTCG-1, AAACCCAAGAAGCTGC-1, AAACCCAAGAAGCTGT-1, AAACCCAAGAAGCTTG-1, AAACCCAAGAAGGACG-1, AAACCCAAGAAGGAGC-1, AAACCCAAGAAGGATG-1, AAACCCAAGAAGGCTC-1, AAACCCAAGAAGGCTT-1, AAACCCAAGAAGGGAT-1, AAACCCAAGAAGGTAG-1, AAACCCAAGAAGGTGG-1, AAACCCAAGAAGTAAG-1, AAACCCAAGAAGTATC-1, AAACCCAAGAAGTCAT-1, AAACCCAAGAAGTCCG-1, AAACCCAAGAAGTCGT-1, AAACCCAAGAAGTCTA-1, AAACCCAAGAAGTCTG-1, AAACCCAAGAAGTGGG-1, AAACCCAAGAAGTGTC-1, AAACCCAAGAAGTGTT-1, AAACCCAAGAAGTTGC-1, AAACCCAAGAATAACC-1, AAACCCAAGAATACAC-1, AAACCCAAGAATAGTC-1, AAACCCAAGAATCAGC-1, AAACCCAAGAATCCAG-1, AAACCCAAGAATCCCT-1, AAACCCAAGAATCCGA-1, AAACCCAAGAATCGAT-1, AAACCCAAGAATCGCG-1, AAACCCAAGAATCGGT-1, AAACCCAAGAATCGTA-1, ...]\n",
       "\n",
       "[6794880 rows x 0 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1298.0"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.max(adata.X[:100000, :])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "162.0"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.max(adata.X[:500, :])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As such, even the values of the count matrix are float, they are still 'raw'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annotate</th>\n",
       "      <th>nReads</th>\n",
       "      <th>nUMI</th>\n",
       "      <th>nGene</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>emptydrop</th>\n",
       "      <th>doublet</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HTAPP-312-SMP-902_fresh-C4-T2_channel1-AAAGATGAGACATAAC</th>\n",
       "      <td>Neuroendocrine</td>\n",
       "      <td>39239</td>\n",
       "      <td>2267</td>\n",
       "      <td>1192</td>\n",
       "      <td>0.059550</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HTAPP-312-SMP-902_fresh-C4-T2_channel1-AAAGATGTCTGGAGCC</th>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>21923</td>\n",
       "      <td>1384</td>\n",
       "      <td>770</td>\n",
       "      <td>0.032514</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HTAPP-312-SMP-902_fresh-C4-T2_channel1-AAAGCAATCTAACCGA</th>\n",
       "      <td>T cell</td>\n",
       "      <td>31269</td>\n",
       "      <td>1780</td>\n",
       "      <td>757</td>\n",
       "      <td>0.024157</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HTAPP-312-SMP-902_fresh-C4-T2_channel1-AAAGTAGAGCTAAACA</th>\n",
       "      <td>Neuroendocrine</td>\n",
       "      <td>59927</td>\n",
       "      <td>3414</td>\n",
       "      <td>1755</td>\n",
       "      <td>0.067955</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HTAPP-312-SMP-902_fresh-C4-T2_channel1-AAAGTAGCAGGAATCG</th>\n",
       "      <td>Neuroendocrine</td>\n",
       "      <td>26150</td>\n",
       "      <td>1508</td>\n",
       "      <td>1008</td>\n",
       "      <td>0.076260</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HTAPP-312-SMP-902_fresh-C4-T2_channel1-TTTGCGCAGTTCCACA</th>\n",
       "      <td>Neuroendocrine</td>\n",
       "      <td>53458</td>\n",
       "      <td>3144</td>\n",
       "      <td>1679</td>\n",
       "      <td>0.034669</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HTAPP-312-SMP-902_fresh-C4-T2_channel1-TTTGCGCCACAGACAG</th>\n",
       "      <td>Neuroendocrine</td>\n",
       "      <td>67111</td>\n",
       "      <td>3874</td>\n",
       "      <td>1787</td>\n",
       "      <td>0.035880</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HTAPP-312-SMP-902_fresh-C4-T2_channel1-TTTGGTTTCATAAAGG</th>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>28516</td>\n",
       "      <td>1809</td>\n",
       "      <td>802</td>\n",
       "      <td>0.039248</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HTAPP-312-SMP-902_fresh-C4-T2_channel1-TTTGTCAAGTTCCACA</th>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>34314</td>\n",
       "      <td>2079</td>\n",
       "      <td>908</td>\n",
       "      <td>0.035113</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HTAPP-312-SMP-902_fresh-C4-T2_channel1-TTTGTCAGTCCAGTAT</th>\n",
       "      <td>Endothelial cell</td>\n",
       "      <td>48448</td>\n",
       "      <td>3543</td>\n",
       "      <td>1500</td>\n",
       "      <td>0.030483</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>786 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                            annotate  nReads  \\\n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-AAAGATGA...    Neuroendocrine   39239   \n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-AAAGATGT...        Fibroblast   21923   \n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-AAAGCAAT...            T cell   31269   \n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-AAAGTAGA...    Neuroendocrine   59927   \n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-AAAGTAGC...    Neuroendocrine   26150   \n",
       "...                                                              ...     ...   \n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-TTTGCGCA...    Neuroendocrine   53458   \n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-TTTGCGCC...    Neuroendocrine   67111   \n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-TTTGGTTT...        Fibroblast   28516   \n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-TTTGTCAA...        Fibroblast   34314   \n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-TTTGTCAG...  Endothelial cell   48448   \n",
       "\n",
       "                                                    nUMI  nGene  percent_mito  \\\n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-AAAGATGA...  2267   1192      0.059550   \n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-AAAGATGT...  1384    770      0.032514   \n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-AAAGCAAT...  1780    757      0.024157   \n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-AAAGTAGA...  3414   1755      0.067955   \n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-AAAGTAGC...  1508   1008      0.076260   \n",
       "...                                                  ...    ...           ...   \n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-TTTGCGCA...  3144   1679      0.034669   \n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-TTTGCGCC...  3874   1787      0.035880   \n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-TTTGGTTT...  1809    802      0.039248   \n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-TTTGTCAA...  2079    908      0.035113   \n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-TTTGTCAG...  3543   1500      0.030483   \n",
       "\n",
       "                                                    emptydrop  doublet  \n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-AAAGATGA...      False    False  \n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-AAAGATGT...      False    False  \n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-AAAGCAAT...      False    False  \n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-AAAGTAGA...      False    False  \n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-AAAGTAGC...      False    False  \n",
       "...                                                       ...      ...  \n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-TTTGCGCA...      False    False  \n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-TTTGCGCC...      False    False  \n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-TTTGGTTT...      False    False  \n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-TTTGTCAA...      False    False  \n",
       "HTAPP-312-SMP-902_fresh-C4-T2_channel1-TTTGTCAG...      False    False  \n",
       "\n",
       "[786 rows x 7 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# inspect the metadata\n",
    "meta_path = data_directory / 'GSM4186962_metadata_HTAPP-312-SMP-902_fresh-C4-T2_channel1.csv'\n",
    "meta = pd.read_csv(meta_path, sep=',', index_col=0)\n",
    "\n",
    "meta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So we can use the metadata to extract filtered cells from the raw h5 files."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. AnnData object of each sample"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Neuroblastoma (NB): HTAPP-312-SMP-901, HTAPP-312-SMP-902, HTAPP-656-SMP-3481, HTAPP-244-SMP-451 (nuclei), HTAPP-656-SMP-3481 (nuclei)\n",
    "\n",
    "Glioblastoma (GB): HTAPP-443-SMP-5491\n",
    "\n",
    "Sarcoma: HTAPP-951-SMP-4652 (nuclei), HTAPP-975-SMP-4771 (nuclei)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the metadata from Figure 1b\n",
    "cancer_tissue_dict = {\n",
    "    'HTAPP-312-SMP-901': ['neuroblastoma', 'neuroendocrine'],\n",
    "    'HTAPP-312-SMP-902': ['neuroblastoma', 'neuroendocrine'],\n",
    "    'HTAPP-656-SMP-3481': ['neuroblastoma', 'neuroendocrine'],\n",
    "    'HTAPP-244-SMP-451': ['neuroblastoma', 'neuroendocrine'],\n",
    "    'HTAPP-443-SMP-5491': ['glioblastoma', 'neuronal'],\n",
    "    'HTAPP-951-SMP-4652': ['sarcoma', 'mesenchymal'],\n",
    "    'HTAPP-975-SMP-4771': ['sarcoma', 'mesenchymal'],\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GSM4186961_HTAPP-312-SMP-901_fresh-T1_channel1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/s4543064/miniforge3/envs/scvi_env/lib/python3.11/site-packages/anndata/_core/anndata.py:1820: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4369, 33694)\n",
      "AnnData object with n_obs × n_vars = 4369 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'cell_type_from_paper', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "GSM4186963_HTAPP-656-SMP-3481_fresh-T1_channel1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/s4543064/miniforge3/envs/scvi_env/lib/python3.11/site-packages/anndata/_core/anndata.py:1820: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3449, 33694)\n",
      "AnnData object with n_obs × n_vars = 3449 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'cell_type_from_paper', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "GSM4186982_HTAPP-443-SMP-5491_CST_channel1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/s4543064/miniforge3/envs/scvi_env/lib/python3.11/site-packages/anndata/_core/anndata.py:1820: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/home/s4543064/miniforge3/envs/scvi_env/lib/python3.11/site-packages/anndata/_core/anndata.py:1820: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3967, 33538)\n",
      "AnnData object with n_obs × n_vars = 3967 × 33538\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'cell_type_from_paper', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "GSM4186994_HTAPP-951-SMP-4652_CST-V3_channel1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/s4543064/miniforge3/envs/scvi_env/lib/python3.11/site-packages/anndata/_core/anndata.py:1820: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/home/s4543064/miniforge3/envs/scvi_env/lib/python3.11/site-packages/anndata/_core/anndata.py:1820: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(7858, 33538)\n",
      "AnnData object with n_obs × n_vars = 7858 × 33538\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'cell_type_from_paper', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "GSM4186993_HTAPP-951-SMP-4652_TST-V3_channel1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/s4543064/miniforge3/envs/scvi_env/lib/python3.11/site-packages/anndata/_core/anndata.py:1820: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/home/s4543064/miniforge3/envs/scvi_env/lib/python3.11/site-packages/anndata/_core/anndata.py:1820: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4458, 33538)\n",
      "AnnData object with n_obs × n_vars = 4458 × 33538\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'cell_type_from_paper', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "GSM4186992_HTAPP-951-SMP-4652_TST-V2_channel1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/s4543064/miniforge3/envs/scvi_env/lib/python3.11/site-packages/anndata/_core/anndata.py:1820: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/home/s4543064/miniforge3/envs/scvi_env/lib/python3.11/site-packages/anndata/_core/anndata.py:1820: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3992, 33694)\n",
      "AnnData object with n_obs × n_vars = 3992 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'cell_type_from_paper', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "GSM4186967_HTAPP-244-SMP-451_NST_channel1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/s4543064/miniforge3/envs/scvi_env/lib/python3.11/site-packages/anndata/_core/anndata.py:1820: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(7531, 33694)\n",
      "AnnData object with n_obs × n_vars = 7531 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'cell_type_from_paper', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "GSM4186969_HTAPP-656-SMP-3481_TST_channel1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/s4543064/miniforge3/envs/scvi_env/lib/python3.11/site-packages/anndata/_core/anndata.py:1820: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(7810, 33694)\n",
      "AnnData object with n_obs × n_vars = 7810 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'cell_type_from_paper', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "GSM4186966_HTAPP-244-SMP-451_EZ_channel1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/s4543064/miniforge3/envs/scvi_env/lib/python3.11/site-packages/anndata/_core/anndata.py:1820: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(7896, 33694)\n",
      "AnnData object with n_obs × n_vars = 7896 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'cell_type_from_paper', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "GSM4186965_HTAPP-244-SMP-451_CST_channel1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/s4543064/miniforge3/envs/scvi_env/lib/python3.11/site-packages/anndata/_core/anndata.py:1820: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(6157, 33694)\n",
      "AnnData object with n_obs × n_vars = 6157 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'cell_type_from_paper', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "GSM4186968_HTAPP-244-SMP-451_TST_channel1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/s4543064/miniforge3/envs/scvi_env/lib/python3.11/site-packages/anndata/_core/anndata.py:1820: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(7415, 33694)\n",
      "AnnData object with n_obs × n_vars = 7415 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'cell_type_from_paper', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "GSM4186995_HTAPP-975-SMP-4771_TST-V3_channel1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/s4543064/miniforge3/envs/scvi_env/lib/python3.11/site-packages/anndata/_core/anndata.py:1820: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/home/s4543064/miniforge3/envs/scvi_env/lib/python3.11/site-packages/anndata/_core/anndata.py:1820: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4317, 33538)\n",
      "AnnData object with n_obs × n_vars = 4317 × 33538\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'cell_type_from_paper', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "GSM4186962_HTAPP-312-SMP-902_fresh-C4-T2_channel1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/s4543064/miniforge3/envs/scvi_env/lib/python3.11/site-packages/anndata/_core/anndata.py:1820: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(786, 33694)\n",
      "AnnData object with n_obs × n_vars = 786 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'cell_type_from_paper', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n"
     ]
    }
   ],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "# Specify directory paths\n",
    "data_directory = Path('/scratch/user/s4543064/xiaohan-john-project/data/GSE140819')\n",
    "write_directory = Path('/scratch/user/s4543064/xiaohan-john-project/write/GSE140819')\n",
    "\n",
    "for sample in data_directory.iterdir():\n",
    "    if '.h5' in sample.name: # eg: sample = GSM4186961_HTAPP-312-SMP-901_fresh-T1_channel1_raw_gene_bc_matrices_h5.h5\n",
    "        # Get the gsm and patient id\n",
    "        gsm = sample.stem.split(\"_\")[0] # gsm = GSM4186961\n",
    "        gsm_patient_id = sample.stem.split(\"_raw_\")[0] # gsm_patient_id = GSM4186961_HTAPP-312-SMP-901_fresh-T1_channel1\n",
    "        print(gsm_patient_id)\n",
    "        patient_id = gsm_patient_id.split('_')[1] # patient_id = HTAPP-312-SMP-901\n",
    "        meta_file = gsm + '_metadata_' + gsm_patient_id.split(gsm + '_')[1] + '.csv'\n",
    "        meta_path = data_directory / meta_file\n",
    "        \n",
    "        meta = pd.read_csv(meta_path, sep=',', index_col=0)\n",
    "        cells = meta.index.tolist() # ['HTAPP-312-SMP-901_fresh-T1_channel1-AAACCTGAGACGCAAC]\n",
    "        # print(cells)\n",
    "        cell_barcodes = [cell.split('channel1')[1][1:] + '-1' for cell in cells] # ['AAACCTGAGACGCAAC-1']\n",
    "        # print(cell_barcodes)\n",
    "        # Reset the index of the metadata to match the one in the anndata\n",
    "        meta.reset_index(drop=True, inplace=True)\n",
    "        meta.index = cell_barcodes\n",
    "\n",
    "        adata = sc.read_10x_h5(sample)\n",
    "        adata.var_names_make_unique()\n",
    "        adata = adata[adata.obs.index.isin(cell_barcodes)].copy()\n",
    "        print(adata.shape)\n",
    "\n",
    "        adata.obs['cancer_type'] = cancer_tissue_dict[patient_id][0]\n",
    "        adata.obs['dataset'] = 'GSE140819'\n",
    "        adata.obs['tissue'] = cancer_tissue_dict[patient_id][1]\n",
    "        adata.obs['cell_type_from_paper'] = meta['annotate']\n",
    "        adata.obs['sample_barcode'] = gsm_patient_id\n",
    "        adata.obs['uni_barcode'] = adata.obs['dataset'] + '_' + adata.obs.index.astype(str)\n",
    "        adata.obs.set_index(\"uni_barcode\", drop=False, inplace=True)\n",
    "        print(adata)\n",
    "\n",
    "        # save the anndata object\n",
    "        sample_h5ad = gsm_patient_id + '_uni.h5ad'\n",
    "        output_path = write_directory / sample_h5ad\n",
    "        adata.write_h5ad(output_path, compression=\"gzip\")\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RP11-34P13.3</th>\n",
       "      <td>ENSG00000243485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>ENSG00000237613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>ENSG00000186092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-34P13.7</th>\n",
       "      <td>ENSG00000238009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-34P13.8</th>\n",
       "      <td>ENSG00000239945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC233755.2</th>\n",
       "      <td>ENSG00000277856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC233755.1</th>\n",
       "      <td>ENSG00000275063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC240274.1</th>\n",
       "      <td>ENSG00000271254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC213203.1</th>\n",
       "      <td>ENSG00000277475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM231B</th>\n",
       "      <td>ENSG00000268674</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33694 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     gene_ids\n",
       "RP11-34P13.3  ENSG00000243485\n",
       "FAM138A       ENSG00000237613\n",
       "OR4F5         ENSG00000186092\n",
       "RP11-34P13.7  ENSG00000238009\n",
       "RP11-34P13.8  ENSG00000239945\n",
       "...                       ...\n",
       "AC233755.2    ENSG00000277856\n",
       "AC233755.1    ENSG00000275063\n",
       "AC240274.1    ENSG00000271254\n",
       "AC213203.1    ENSG00000277475\n",
       "FAM231B       ENSG00000268674\n",
       "\n",
       "[33694 rows x 1 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cancer_type</th>\n",
       "      <th>dataset</th>\n",
       "      <th>tissue</th>\n",
       "      <th>cell_type_from_paper</th>\n",
       "      <th>sample_barcode</th>\n",
       "      <th>uni_barcode</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>uni_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSE140819_AAAGATGAGACATAAC-1</th>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>GSE140819</td>\n",
       "      <td>neuroendocrine</td>\n",
       "      <td>Neuroendocrine</td>\n",
       "      <td>GSM4186962_HTAPP-312-SMP-902_fresh-C4-T2_channel1</td>\n",
       "      <td>GSE140819_AAAGATGAGACATAAC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE140819_AAAGATGTCTGGAGCC-1</th>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>GSE140819</td>\n",
       "      <td>neuroendocrine</td>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>GSM4186962_HTAPP-312-SMP-902_fresh-C4-T2_channel1</td>\n",
       "      <td>GSE140819_AAAGATGTCTGGAGCC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE140819_AAAGCAATCTAACCGA-1</th>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>GSE140819</td>\n",
       "      <td>neuroendocrine</td>\n",
       "      <td>T cell</td>\n",
       "      <td>GSM4186962_HTAPP-312-SMP-902_fresh-C4-T2_channel1</td>\n",
       "      <td>GSE140819_AAAGCAATCTAACCGA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE140819_AAAGTAGAGCTAAACA-1</th>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>GSE140819</td>\n",
       "      <td>neuroendocrine</td>\n",
       "      <td>Neuroendocrine</td>\n",
       "      <td>GSM4186962_HTAPP-312-SMP-902_fresh-C4-T2_channel1</td>\n",
       "      <td>GSE140819_AAAGTAGAGCTAAACA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE140819_AAAGTAGCAGGAATCG-1</th>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>GSE140819</td>\n",
       "      <td>neuroendocrine</td>\n",
       "      <td>Neuroendocrine</td>\n",
       "      <td>GSM4186962_HTAPP-312-SMP-902_fresh-C4-T2_channel1</td>\n",
       "      <td>GSE140819_AAAGTAGCAGGAATCG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE140819_TTTGCGCAGTTCCACA-1</th>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>GSE140819</td>\n",
       "      <td>neuroendocrine</td>\n",
       "      <td>Neuroendocrine</td>\n",
       "      <td>GSM4186962_HTAPP-312-SMP-902_fresh-C4-T2_channel1</td>\n",
       "      <td>GSE140819_TTTGCGCAGTTCCACA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE140819_TTTGCGCCACAGACAG-1</th>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>GSE140819</td>\n",
       "      <td>neuroendocrine</td>\n",
       "      <td>Neuroendocrine</td>\n",
       "      <td>GSM4186962_HTAPP-312-SMP-902_fresh-C4-T2_channel1</td>\n",
       "      <td>GSE140819_TTTGCGCCACAGACAG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE140819_TTTGGTTTCATAAAGG-1</th>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>GSE140819</td>\n",
       "      <td>neuroendocrine</td>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>GSM4186962_HTAPP-312-SMP-902_fresh-C4-T2_channel1</td>\n",
       "      <td>GSE140819_TTTGGTTTCATAAAGG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE140819_TTTGTCAAGTTCCACA-1</th>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>GSE140819</td>\n",
       "      <td>neuroendocrine</td>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>GSM4186962_HTAPP-312-SMP-902_fresh-C4-T2_channel1</td>\n",
       "      <td>GSE140819_TTTGTCAAGTTCCACA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE140819_TTTGTCAGTCCAGTAT-1</th>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>GSE140819</td>\n",
       "      <td>neuroendocrine</td>\n",
       "      <td>Endothelial cell</td>\n",
       "      <td>GSM4186962_HTAPP-312-SMP-902_fresh-C4-T2_channel1</td>\n",
       "      <td>GSE140819_TTTGTCAGTCCAGTAT-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>786 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                cancer_type    dataset          tissue  \\\n",
       "uni_barcode                                                              \n",
       "GSE140819_AAAGATGAGACATAAC-1  neuroblastoma  GSE140819  neuroendocrine   \n",
       "GSE140819_AAAGATGTCTGGAGCC-1  neuroblastoma  GSE140819  neuroendocrine   \n",
       "GSE140819_AAAGCAATCTAACCGA-1  neuroblastoma  GSE140819  neuroendocrine   \n",
       "GSE140819_AAAGTAGAGCTAAACA-1  neuroblastoma  GSE140819  neuroendocrine   \n",
       "GSE140819_AAAGTAGCAGGAATCG-1  neuroblastoma  GSE140819  neuroendocrine   \n",
       "...                                     ...        ...             ...   \n",
       "GSE140819_TTTGCGCAGTTCCACA-1  neuroblastoma  GSE140819  neuroendocrine   \n",
       "GSE140819_TTTGCGCCACAGACAG-1  neuroblastoma  GSE140819  neuroendocrine   \n",
       "GSE140819_TTTGGTTTCATAAAGG-1  neuroblastoma  GSE140819  neuroendocrine   \n",
       "GSE140819_TTTGTCAAGTTCCACA-1  neuroblastoma  GSE140819  neuroendocrine   \n",
       "GSE140819_TTTGTCAGTCCAGTAT-1  neuroblastoma  GSE140819  neuroendocrine   \n",
       "\n",
       "                             cell_type_from_paper  \\\n",
       "uni_barcode                                         \n",
       "GSE140819_AAAGATGAGACATAAC-1       Neuroendocrine   \n",
       "GSE140819_AAAGATGTCTGGAGCC-1           Fibroblast   \n",
       "GSE140819_AAAGCAATCTAACCGA-1               T cell   \n",
       "GSE140819_AAAGTAGAGCTAAACA-1       Neuroendocrine   \n",
       "GSE140819_AAAGTAGCAGGAATCG-1       Neuroendocrine   \n",
       "...                                           ...   \n",
       "GSE140819_TTTGCGCAGTTCCACA-1       Neuroendocrine   \n",
       "GSE140819_TTTGCGCCACAGACAG-1       Neuroendocrine   \n",
       "GSE140819_TTTGGTTTCATAAAGG-1           Fibroblast   \n",
       "GSE140819_TTTGTCAAGTTCCACA-1           Fibroblast   \n",
       "GSE140819_TTTGTCAGTCCAGTAT-1     Endothelial cell   \n",
       "\n",
       "                                                                 sample_barcode  \\\n",
       "uni_barcode                                                                       \n",
       "GSE140819_AAAGATGAGACATAAC-1  GSM4186962_HTAPP-312-SMP-902_fresh-C4-T2_channel1   \n",
       "GSE140819_AAAGATGTCTGGAGCC-1  GSM4186962_HTAPP-312-SMP-902_fresh-C4-T2_channel1   \n",
       "GSE140819_AAAGCAATCTAACCGA-1  GSM4186962_HTAPP-312-SMP-902_fresh-C4-T2_channel1   \n",
       "GSE140819_AAAGTAGAGCTAAACA-1  GSM4186962_HTAPP-312-SMP-902_fresh-C4-T2_channel1   \n",
       "GSE140819_AAAGTAGCAGGAATCG-1  GSM4186962_HTAPP-312-SMP-902_fresh-C4-T2_channel1   \n",
       "...                                                                         ...   \n",
       "GSE140819_TTTGCGCAGTTCCACA-1  GSM4186962_HTAPP-312-SMP-902_fresh-C4-T2_channel1   \n",
       "GSE140819_TTTGCGCCACAGACAG-1  GSM4186962_HTAPP-312-SMP-902_fresh-C4-T2_channel1   \n",
       "GSE140819_TTTGGTTTCATAAAGG-1  GSM4186962_HTAPP-312-SMP-902_fresh-C4-T2_channel1   \n",
       "GSE140819_TTTGTCAAGTTCCACA-1  GSM4186962_HTAPP-312-SMP-902_fresh-C4-T2_channel1   \n",
       "GSE140819_TTTGTCAGTCCAGTAT-1  GSM4186962_HTAPP-312-SMP-902_fresh-C4-T2_channel1   \n",
       "\n",
       "                                               uni_barcode  \n",
       "uni_barcode                                                 \n",
       "GSE140819_AAAGATGAGACATAAC-1  GSE140819_AAAGATGAGACATAAC-1  \n",
       "GSE140819_AAAGATGTCTGGAGCC-1  GSE140819_AAAGATGTCTGGAGCC-1  \n",
       "GSE140819_AAAGCAATCTAACCGA-1  GSE140819_AAAGCAATCTAACCGA-1  \n",
       "GSE140819_AAAGTAGAGCTAAACA-1  GSE140819_AAAGTAGAGCTAAACA-1  \n",
       "GSE140819_AAAGTAGCAGGAATCG-1  GSE140819_AAAGTAGCAGGAATCG-1  \n",
       "...                                                    ...  \n",
       "GSE140819_TTTGCGCAGTTCCACA-1  GSE140819_TTTGCGCAGTTCCACA-1  \n",
       "GSE140819_TTTGCGCCACAGACAG-1  GSE140819_TTTGCGCCACAGACAG-1  \n",
       "GSE140819_TTTGGTTTCATAAAGG-1  GSE140819_TTTGGTTTCATAAAGG-1  \n",
       "GSE140819_TTTGTCAAGTTCCACA-1  GSE140819_TTTGTCAAGTTCCACA-1  \n",
       "GSE140819_TTTGTCAGTCCAGTAT-1  GSE140819_TTTGTCAGTCCAGTAT-1  \n",
       "\n",
       "[786 rows x 6 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Confirmation of created AnnData objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 7415 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'cell_type_from_paper', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 3992 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'cell_type_from_paper', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "AnnData object with n_obs × n_vars = 7896 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'cell_type_from_paper', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 786 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'cell_type_from_paper', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 7810 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'cell_type_from_paper', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 3967 × 33538\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'cell_type_from_paper', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "AnnData object with n_obs × n_vars = 7858 × 33538\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'cell_type_from_paper', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "AnnData object with n_obs × n_vars = 3449 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'cell_type_from_paper', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 7531 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'cell_type_from_paper', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 4458 × 33538\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'cell_type_from_paper', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n",
      "AnnData object with n_obs × n_vars = 4369 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'cell_type_from_paper', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 6157 × 33694\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'cell_type_from_paper', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids'\n",
      "AnnData object with n_obs × n_vars = 4317 × 33538\n",
      "    obs: 'cancer_type', 'dataset', 'tissue', 'cell_type_from_paper', 'sample_barcode', 'uni_barcode'\n",
      "    var: 'gene_ids', 'feature_types', 'genome'\n"
     ]
    }
   ],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "# Specify directory paths\n",
    "write_directory = Path('/scratch/user/s4543064/xiaohan-john-project/write/GSE140819')\n",
    "\n",
    "# Loop through all files in the directory\n",
    "for file in write_directory.iterdir():\n",
    "    if '_uni.h5ad' in file.name:\n",
    "        sample = anndata.read_h5ad(file)\n",
    "        print(sample)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>genome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.3</th>\n",
       "      <td>ENSG00000239945</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC233755.2</th>\n",
       "      <td>ENSG00000277856</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC233755.1</th>\n",
       "      <td>ENSG00000275063</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC240274.1</th>\n",
       "      <td>ENSG00000271254</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC213203.1</th>\n",
       "      <td>ENSG00000277475</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM231C</th>\n",
       "      <td>ENSG00000268674</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-3.0.0_premrna</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33538 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    gene_ids    feature_types                genome\n",
       "MIR1302-2HG  ENSG00000243485  Gene Expression  GRCh38-3.0.0_premrna\n",
       "FAM138A      ENSG00000237613  Gene Expression  GRCh38-3.0.0_premrna\n",
       "OR4F5        ENSG00000186092  Gene Expression  GRCh38-3.0.0_premrna\n",
       "AL627309.1   ENSG00000238009  Gene Expression  GRCh38-3.0.0_premrna\n",
       "AL627309.3   ENSG00000239945  Gene Expression  GRCh38-3.0.0_premrna\n",
       "...                      ...              ...                   ...\n",
       "AC233755.2   ENSG00000277856  Gene Expression  GRCh38-3.0.0_premrna\n",
       "AC233755.1   ENSG00000275063  Gene Expression  GRCh38-3.0.0_premrna\n",
       "AC240274.1   ENSG00000271254  Gene Expression  GRCh38-3.0.0_premrna\n",
       "AC213203.1   ENSG00000277475  Gene Expression  GRCh38-3.0.0_premrna\n",
       "FAM231C      ENSG00000268674  Gene Expression  GRCh38-3.0.0_premrna\n",
       "\n",
       "[33538 rows x 3 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cancer_type</th>\n",
       "      <th>dataset</th>\n",
       "      <th>tissue</th>\n",
       "      <th>cell_type_from_paper</th>\n",
       "      <th>sample_barcode</th>\n",
       "      <th>uni_barcode</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>uni_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSE140819_AAACCCACATTGTGCA-1</th>\n",
       "      <td>sarcoma</td>\n",
       "      <td>GSE140819</td>\n",
       "      <td>mesenchymal</td>\n",
       "      <td>Endothelial cell</td>\n",
       "      <td>GSM4186995_HTAPP-975-SMP-4771_TST-V3_channel1</td>\n",
       "      <td>GSE140819_AAACCCACATTGTGCA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE140819_AAACCCAGTGTTCATG-1</th>\n",
       "      <td>sarcoma</td>\n",
       "      <td>GSE140819</td>\n",
       "      <td>mesenchymal</td>\n",
       "      <td>Macrophage 1</td>\n",
       "      <td>GSM4186995_HTAPP-975-SMP-4771_TST-V3_channel1</td>\n",
       "      <td>GSE140819_AAACCCAGTGTTCATG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE140819_AAACCCATCATTCGTT-1</th>\n",
       "      <td>sarcoma</td>\n",
       "      <td>GSE140819</td>\n",
       "      <td>mesenchymal</td>\n",
       "      <td>Endothelial cell</td>\n",
       "      <td>GSM4186995_HTAPP-975-SMP-4771_TST-V3_channel1</td>\n",
       "      <td>GSE140819_AAACCCATCATTCGTT-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE140819_AAACCCATCCTGATAG-1</th>\n",
       "      <td>sarcoma</td>\n",
       "      <td>GSE140819</td>\n",
       "      <td>mesenchymal</td>\n",
       "      <td>Endothelial cell</td>\n",
       "      <td>GSM4186995_HTAPP-975-SMP-4771_TST-V3_channel1</td>\n",
       "      <td>GSE140819_AAACCCATCCTGATAG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE140819_AAACGAAAGCTTTCTT-1</th>\n",
       "      <td>sarcoma</td>\n",
       "      <td>GSE140819</td>\n",
       "      <td>mesenchymal</td>\n",
       "      <td>Chondrocyte</td>\n",
       "      <td>GSM4186995_HTAPP-975-SMP-4771_TST-V3_channel1</td>\n",
       "      <td>GSE140819_AAACGAAAGCTTTCTT-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE140819_TTTGGTTCATGGGATG-1</th>\n",
       "      <td>sarcoma</td>\n",
       "      <td>GSE140819</td>\n",
       "      <td>mesenchymal</td>\n",
       "      <td>Endothelial cell</td>\n",
       "      <td>GSM4186995_HTAPP-975-SMP-4771_TST-V3_channel1</td>\n",
       "      <td>GSE140819_TTTGGTTCATGGGATG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE140819_TTTGTTGGTACCTTCC-1</th>\n",
       "      <td>sarcoma</td>\n",
       "      <td>GSE140819</td>\n",
       "      <td>mesenchymal</td>\n",
       "      <td>Endothelial cell</td>\n",
       "      <td>GSM4186995_HTAPP-975-SMP-4771_TST-V3_channel1</td>\n",
       "      <td>GSE140819_TTTGTTGGTACCTTCC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE140819_TTTGTTGGTCCACATA-1</th>\n",
       "      <td>sarcoma</td>\n",
       "      <td>GSE140819</td>\n",
       "      <td>mesenchymal</td>\n",
       "      <td>Chondrocyte</td>\n",
       "      <td>GSM4186995_HTAPP-975-SMP-4771_TST-V3_channel1</td>\n",
       "      <td>GSE140819_TTTGTTGGTCCACATA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE140819_TTTGTTGGTTTCGTAG-1</th>\n",
       "      <td>sarcoma</td>\n",
       "      <td>GSE140819</td>\n",
       "      <td>mesenchymal</td>\n",
       "      <td>Endothelial cell</td>\n",
       "      <td>GSM4186995_HTAPP-975-SMP-4771_TST-V3_channel1</td>\n",
       "      <td>GSE140819_TTTGTTGGTTTCGTAG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSE140819_TTTGTTGTCTGAGCAT-1</th>\n",
       "      <td>sarcoma</td>\n",
       "      <td>GSE140819</td>\n",
       "      <td>mesenchymal</td>\n",
       "      <td>Fibroblast?</td>\n",
       "      <td>GSM4186995_HTAPP-975-SMP-4771_TST-V3_channel1</td>\n",
       "      <td>GSE140819_TTTGTTGTCTGAGCAT-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4317 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             cancer_type    dataset       tissue  \\\n",
       "uni_barcode                                                        \n",
       "GSE140819_AAACCCACATTGTGCA-1     sarcoma  GSE140819  mesenchymal   \n",
       "GSE140819_AAACCCAGTGTTCATG-1     sarcoma  GSE140819  mesenchymal   \n",
       "GSE140819_AAACCCATCATTCGTT-1     sarcoma  GSE140819  mesenchymal   \n",
       "GSE140819_AAACCCATCCTGATAG-1     sarcoma  GSE140819  mesenchymal   \n",
       "GSE140819_AAACGAAAGCTTTCTT-1     sarcoma  GSE140819  mesenchymal   \n",
       "...                                  ...        ...          ...   \n",
       "GSE140819_TTTGGTTCATGGGATG-1     sarcoma  GSE140819  mesenchymal   \n",
       "GSE140819_TTTGTTGGTACCTTCC-1     sarcoma  GSE140819  mesenchymal   \n",
       "GSE140819_TTTGTTGGTCCACATA-1     sarcoma  GSE140819  mesenchymal   \n",
       "GSE140819_TTTGTTGGTTTCGTAG-1     sarcoma  GSE140819  mesenchymal   \n",
       "GSE140819_TTTGTTGTCTGAGCAT-1     sarcoma  GSE140819  mesenchymal   \n",
       "\n",
       "                             cell_type_from_paper  \\\n",
       "uni_barcode                                         \n",
       "GSE140819_AAACCCACATTGTGCA-1     Endothelial cell   \n",
       "GSE140819_AAACCCAGTGTTCATG-1         Macrophage 1   \n",
       "GSE140819_AAACCCATCATTCGTT-1     Endothelial cell   \n",
       "GSE140819_AAACCCATCCTGATAG-1     Endothelial cell   \n",
       "GSE140819_AAACGAAAGCTTTCTT-1          Chondrocyte   \n",
       "...                                           ...   \n",
       "GSE140819_TTTGGTTCATGGGATG-1     Endothelial cell   \n",
       "GSE140819_TTTGTTGGTACCTTCC-1     Endothelial cell   \n",
       "GSE140819_TTTGTTGGTCCACATA-1          Chondrocyte   \n",
       "GSE140819_TTTGTTGGTTTCGTAG-1     Endothelial cell   \n",
       "GSE140819_TTTGTTGTCTGAGCAT-1          Fibroblast?   \n",
       "\n",
       "                                                             sample_barcode  \\\n",
       "uni_barcode                                                                   \n",
       "GSE140819_AAACCCACATTGTGCA-1  GSM4186995_HTAPP-975-SMP-4771_TST-V3_channel1   \n",
       "GSE140819_AAACCCAGTGTTCATG-1  GSM4186995_HTAPP-975-SMP-4771_TST-V3_channel1   \n",
       "GSE140819_AAACCCATCATTCGTT-1  GSM4186995_HTAPP-975-SMP-4771_TST-V3_channel1   \n",
       "GSE140819_AAACCCATCCTGATAG-1  GSM4186995_HTAPP-975-SMP-4771_TST-V3_channel1   \n",
       "GSE140819_AAACGAAAGCTTTCTT-1  GSM4186995_HTAPP-975-SMP-4771_TST-V3_channel1   \n",
       "...                                                                     ...   \n",
       "GSE140819_TTTGGTTCATGGGATG-1  GSM4186995_HTAPP-975-SMP-4771_TST-V3_channel1   \n",
       "GSE140819_TTTGTTGGTACCTTCC-1  GSM4186995_HTAPP-975-SMP-4771_TST-V3_channel1   \n",
       "GSE140819_TTTGTTGGTCCACATA-1  GSM4186995_HTAPP-975-SMP-4771_TST-V3_channel1   \n",
       "GSE140819_TTTGTTGGTTTCGTAG-1  GSM4186995_HTAPP-975-SMP-4771_TST-V3_channel1   \n",
       "GSE140819_TTTGTTGTCTGAGCAT-1  GSM4186995_HTAPP-975-SMP-4771_TST-V3_channel1   \n",
       "\n",
       "                                               uni_barcode  \n",
       "uni_barcode                                                 \n",
       "GSE140819_AAACCCACATTGTGCA-1  GSE140819_AAACCCACATTGTGCA-1  \n",
       "GSE140819_AAACCCAGTGTTCATG-1  GSE140819_AAACCCAGTGTTCATG-1  \n",
       "GSE140819_AAACCCATCATTCGTT-1  GSE140819_AAACCCATCATTCGTT-1  \n",
       "GSE140819_AAACCCATCCTGATAG-1  GSE140819_AAACCCATCCTGATAG-1  \n",
       "GSE140819_AAACGAAAGCTTTCTT-1  GSE140819_AAACGAAAGCTTTCTT-1  \n",
       "...                                                    ...  \n",
       "GSE140819_TTTGGTTCATGGGATG-1  GSE140819_TTTGGTTCATGGGATG-1  \n",
       "GSE140819_TTTGTTGGTACCTTCC-1  GSE140819_TTTGTTGGTACCTTCC-1  \n",
       "GSE140819_TTTGTTGGTCCACATA-1  GSE140819_TTTGTTGGTCCACATA-1  \n",
       "GSE140819_TTTGTTGGTTTCGTAG-1  GSE140819_TTTGTTGGTTTCGTAG-1  \n",
       "GSE140819_TTTGTTGTCTGAGCAT-1  GSE140819_TTTGTTGTCTGAGCAT-1  \n",
       "\n",
       "[4317 rows x 6 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. Convert AnnData objects to SingleCellExperiment objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'anndata2ri'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[36], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mpathlib\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m Path\n\u001b[0;32m----> 3\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01manndata2ri\u001b[39;00m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mrpy2\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mrobjects\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mrobjects\u001b[39;00m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mrpy2\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mrobjects\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m r\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'anndata2ri'"
     ]
    }
   ],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "import anndata2ri\n",
    "import rpy2.robjects as robjects\n",
    "from rpy2.robjects import r\n",
    "from rpy2.robjects.conversion import localconverter\n",
    "\n",
    "# Specify directory paths\n",
    "write_directory = Path('/scratch/user/s4543064/xiaohan-john-project/write/GSE140819')\n",
    "\n",
    "# Loop through all files in the directory\n",
    "for file in write_directory.iterdir():\n",
    "    sample_name = file.stem\n",
    "    if \"_uni.h5ad\" in file.name:\n",
    "        sample_anndata = anndata.read_h5ad(file)\n",
    "        sample_sce_file = sample_name + \".rds\"\n",
    "\n",
    "        with localconverter(anndata2ri.converter):\n",
    "            sample_sce = anndata2ri.py2rpy(sample_anndata)\n",
    "        \n",
    "        # Save the sce object in .rds file\n",
    "        robjects.globalenv[\"sample_sce\"] = sample_sce\n",
    "        sample_sce_path = write_directory / sample_sce_file\n",
    "        robjects.r(\"saveRDS(sample_sce, file='{}')\".format(sample_sce_path))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "class: SingleCellExperiment \n",
      "dim: 33694 737280 \n",
      "metadata(0):\n",
      "assays(1): X\n",
      "rownames(33694): RP11-34P13.3 FAM138A ... AC213203.1 FAM231B\n",
      "rowData names(1): gene_ids\n",
      "colnames(737280): GSE140819_AAACCTGAGAAACCAT-1\n",
      "  GSE140819_AAACCTGAGAAACCGC-1 ... GSE140819_TTTGTCATCTTTAGTC-1\n",
      "  GSE140819_TTTGTCATCTTTCCTC-1\n",
      "colData names(5): cancer_type dataset tissue sample_barcode uni_barcode\n",
      "reducedDimNames(0):\n",
      "mainExpName: NULL\n",
      "altExpNames(0):\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(sample_sce)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
